miriway.cpp

```diff
@@ -68,24 +71,70 @@ std::map<std::string, ShellCommands::CmdFunctor> const wm_command =
 // Build a list of shell components from "<commands>" values and launch them all after startup
 struct ShellComponents
 {
-    explicit ShellComponents(std::function<void(std::vector<std::string> const& command_line)> launch) :
+    explicit ShellComponents(std::function<pid_t(std::vector<std::string> const& command_line)> launch) :
         launch{std::move(launch)}{}
 
     void populate(std::vector<std::string> const& cmds)
     {
         std::transform(begin(cmds), end(cmds), back_inserter(commands), ExternalClientLauncher::split_command);
     }
 
-    void launch_all() const
+    void populate_auto_restart(std::vector<std::string> const& cmds)
     {
+        std::transform(begin(cmds), end(cmds), back_inserter(auto_restart_commands), ExternalClientLauncher::split_command);
+    }
+
+    static void run_process_restart_thread(
+        bool& is_running,
+        std::vector<std::vector<std::string>> auto_restart_commands,
+        std::vector<pid_t> watched_process_list,
+        std::function<pid_t(std::vector<std::string> const& command_line)> launch)
+    {
+        while (is_running)
+        {
+            for (size_t i = 0; i < watched_process_list.size(); i++)
+            {
+                auto pid = watched_process_list[i];
```

Comment:
```suggestion
            for (auto& pid : watched_process_list)
            {
```

Reply:
I am going to actually use the "process dead" notification that already exists, so this will disappear. Will let you know when it's done

Reply:
Sounds like this should be "draft" then